### DISCUSSION TITLE
x-on:input.stop also prevents $dispatch('input', value)

### DISCUSSION THREAD

**Discussion by mblarsen** [alpinejs/alpine#569] [4. General]
I was writing an inline type editor. My goal is to only emit 'input' after the user hits enter.

```html
<x-input.inline wire:model="name" />
```

and the implementation:

```html
<div
    x-data="{edit: false, value: '', original: ''}"
    x-init="original = value = $refs.input.value"
    wire:ignore
>
    <div
        x-show="!edit"
        x-text="value"
        x-on:click="edit = true; $nextTick(() => $refs.input.focus())"
    >
        {{ $value }}
    </div>


    <x-input.text
        x-cloak
        x-ref="input"
        x-show="edit"
        x-on:keydown.escape="value = original; edit = false"
        x-on:keydown.enter.prevent="$dispatch('input', value); edit = false;"
        x-on:input.stop="value = $event.target.value"
   />
</div>
```

So when I add `.stop` to `x-on:input` I'm no longer able to `$dispatch('input', value)`. If I change dispatched name to something else dispatch does emit.

Am I doing anything wrong? Any suggested workaround? Bug?

**Comment by ryangjchandler** [comment]
I'm not sure what your particular use case is, so it's hard to suggest alternatives. Why is it you only want to dispatch an `input` event when pressing enter?

**Reply by mblarsen** [reply]
When you click the title of a task list (in my case) the list title is replaced with the input field. The user can edit the title but only when the user hits enter the change should be persisted.

The reason why I wanted to use input is because of how Livewire works. E.g. I have `wire:model="name"` which is an event listener for 'input'. If I for instance dispatched 'change' instead I would have to use `wire:model.lazy="name"` which just seems like bad DX for the user of the component. Even worse if I had to wire up all of this manually with another event name.

But do you have any idea why `x-on:input.stop` would prevent the dispatch of the event with the same name?. I wonder if the order matter 🤔️

**Reply by mblarsen** [reply]
It didn't

**Reply by ryangjchandler** [reply]
`.stop` prevents propagation of an event, so it won't bubble up to a parent element. Have you tried using the `.self` modifier instead?

**Reply by mblarsen** [reply]
Yeah I tried `.self` it fits the bill in theory, but the events still bubbles to Livewire's `wire:model="name"`.

**Reply by ryangjchandler** [reply]
Ah, yeah of course. What does the code for the actual component look like? Is there a wrapping div or is just the input field?

**Reply by mblarsen** [reply]
```html
<!-- input.inline -->
@props([
    'id' => false,
    'name' => false,
    'inputClass' => '',
    'inputStyle' => '',
    'value' => '…',
    'persist' => true,
])

<div
    x-data="{edit: false, value: '', original: ''}"
    x-init="original = value = $refs.input.value"
    class="w-full Inline"
    {{ $attributes->except('class') }}
>
    <div wire:ignore>
        <div
            x-show="!edit"
            x-text="value"
            x-on:click="edit = true; $nextTick(() => $refs.input.focus())"
            class="{{ $attributes['class'] }}"
            :class="{'font-bold text-primary': value !== original}"
        >
            {{ $value }}
        </div>

        <x-input.text
            x-cloak
            x-ref="input"
            x-show="edit"
            x-model.stop="value"
            x-on:keydown.escape="value = original; $dispatch('change', value); edit = false"
            x-on:keydown.enter.prevent="$dispatch('change', value); edit = false; if ({{ $persist ? 'true' : 'false' }}) { original = value }"
            :id="$id ?: $name"
            :name="$name"
            :class="$inputClass"
            :style="$inputStyle"
            :value="$value"
            />
    </div>
</div>

<!-- input.text -->

<input
    {{
        $attributes
            ->merge([
                'id' => $attributes['id'] ?? $attributes['name'],
                'type' => $attributes['type'] ?? 'text',
                'autocomplete' => ($attributes['autocomplete'] ?? false) ? 'on' : 'off',
                'class' =>
                        "relative inline-flex items-center justify-start w-full h-12 max-w-full
                         px-3 py-2 text-xl text-black bg-white border-2 border-solid rounded-sm
                         appearance-none sibling placeholder-grey-dark leading-6
                         border-grey-lighter focus:border-primary"
            ])
    }}
/>

<!-- and used like this -->

        <x-input.inline
            class="text-2xl"
            input-class="text-2xl bg-transparent border-none"
            input-style="margin: 0; padding: 0; height: 36px"
            wire:model.lazy="name"
            value="{{ $name }}"
        />
```

The Livewire component just has a public prop `name` in this case.

**Reply by mblarsen** [reply]
Are you also a Livewire user if not maybe it can be a bit messy to look at? :)

**Reply by ryangjchandler** [reply]
Yeah, I'm a full-time Laravel developer and use Livewire on a day-to-day basis.

**Reply by ryangjchandler** [reply]
The `input` event won't be dispatched because `.stop` will prevent event bubbling. It will reach the input itself, but won't go any further up the chain, hence why the Livewire component doesn't react.

I'm finding it hard to see why the `x-on:input` is needed at all, you simply only care about the `keydown.enter`. Instead of updating the value of `value`, just use `$event.target.value` to pull the current text value of the input. That way, you don't need to use the `value` property anymore.

You can keep the `original` property, and when the `keydown.escape` handler runs, just do `$event.target.value = original`.

**Reply by mblarsen** [reply]
Ah, I think I know what happens.

The `$dispatch(event)` is sugar for `target.dispatchEvent(event)` where event in this case is `input`. When the event is dispatched the other handler `x-on:input.stop` catches it and stops propagation. End of story.

**Reply by mblarsen** [reply]
> I'm finding it hard to see why the x-on:input is needed at all,

It is because if I do not stop propagation the native input event from the input field is picked up instead. So leaving `x-on:input` or `x-model` off entirely has the same effect at not using `.stop` and leaving them on.

**Reply by mblarsen** [reply]
But you are right that I don't have to maintain/update it internally except for the convenience of updating the text of the inline editor when it returns to non-edit mode.

**Reply by ryangjchandler** [reply]
> Ah, I think I know what happens.
> 
> The `$dispatch(event)` is sugar for `target.dispatchEvent(event)` where event in this case is `input`. When the event is dispatched the other handler `x-on:input.stop` catches it and stops propagation. End of story.

Yeah, exactly that. There's no need to track the `value` prop, since you have access to that via the input itself. It would be useful if you needed to share the value, but in your case, you don't need to do that.

**Reply by mblarsen** [reply]
Anyway, thanks for taking a look.

I'll try a different approach where using Alpine with custom event on intermediary parent to turn the event from 'custom-input' to 'input'.

So on the input field I'll use `.stop` and then on `.enter` I'll dispatch `custom-input` then on on intermediary parent: `@custom-input="$dispatch('input', $event.detail)"`. Let's see how that goes.

**Reply by ryangjchandler** [reply]
Yeah, if you want to use `.stop` on the input element, you'll need to dispatch the `input` event from outside of it, otherwise it will be stopped by the input.

**Reply by mblarsen** [reply]
So I got it working using the above mentioned approach.

```diff
<div
    x-data="{edit: false, value: '', original: ''}"
    x-init="original = value = $refs.input.value"
>
    <div
        wire:ignore
+       x-on:custom-input="
+           $el.value = $event.detail
+           $el.dispatchEvent(new Event('input', {
+               'bubbles': true,
+               'cancelable': true
+           }))
        "
    >
        ...
        ...
        ...
        <x-input.text
            ...
+           x-model.stop="value" // simply to stop
            ...
            x-on:keydown.enter.prevent="
+               $dispatch('custom-input', value);
                edit = false;"
```

All that to use:

```diff
<x-input.inline
-   wire:model.lazy="name"
+   wire:model="name"
/>
```

**Reply by ryangjchandler** [reply]
Glad you figured this one out.

Out of curiosity, have you tried moving the `x-on:keydown.enter.prevent` to the wrapping `<div>`, then directly dispatching the `input` from there. I think that might remove the need for the `custom-input` event.

**Reply by mblarsen** [reply]
No, that's not working either for the same reason. One approach different from the custom event that would work would be to stop using `.stop` and then manually invoking `$el.stopPropagation()` depending on the key pressed. But that requires some state too, so not better off. I actually prefer the custom event then. Until something better comes along. Thanks for the effort :+1:
